@typedef creates namespace in typescript code

**TypeScript Version:**  nightly (2.4.0-dev.20170608)

**Code**

```ts
namespace N {
    export namespace M {
        export function f(_x: N.M.T): void {}
        export type T = string;
    }

    /** @typedef {string} N.Str */
    export type Str = string;
}
```

**Expected behavior:**

No error.

**Actual behavior:**

```
src/a.ts(3,31): error TS4078: Parameter '_x' of exported function has or is using private name 'N'.
src/a.ts(3,33): error TS2694: Namespace 'N' has no exported member 'M'.
```

(Broken by #15856)

--- a/src/compiler/binder.ts
+++ b/src/compiler/binder.ts
@@ -2138,8 +2138,10 @@ namespace ts {
                 case SyntaxKind.EnumDeclaration:
                     return bindEnumDeclaration(<EnumDeclaration>node);
                 case SyntaxKind.ModuleDeclaration:
-                    return bindModuleDeclaration(<ModuleDeclaration>node);
-
+                    if (node.parent.kind !== ts.SyntaxKind.JSDocTypedefTag || isInJavaScriptFile(node)) {
+                        return bindModuleDeclaration(<ModuleDeclaration>node);
+                    }
+                    return undefined;
                 // Jsx-attributes
                 case SyntaxKind.JsxAttributes:
                     return bindJsxAttributes(<JsxAttributes>node);

--- a/src/compiler/parser.ts
+++ b/src/compiler/parser.ts
@@ -6851,7 +6851,7 @@ namespace ts {
                             jsDocNamespaceNode.flags |= flags;
                             jsDocNamespaceNode.name = typeNameOrNamespaceName;
                             jsDocNamespaceNode.body = parseJSDocTypeNameWithNamespace(NodeFlags.NestedNamespace);
-                            return jsDocNamespaceNode;
+                            return finishNode(jsDocNamespaceNode);
                         }
 
                         if (typeNameOrNamespaceName && flags & NodeFlags.NestedNamespace) {

--- a/tests/baselines/reference/jsdocInTypeScript.errors.txt
+++ b/tests/baselines/reference/jsdocInTypeScript.errors.txt
@@ -3,9 +3,10 @@ tests/cases/compiler/jsdocInTypeScript.ts(23,33): error TS2362: The left-hand si
 tests/cases/compiler/jsdocInTypeScript.ts(25,3): error TS2345: Argument of type '1' is not assignable to parameter of type 'boolean'.
 tests/cases/compiler/jsdocInTypeScript.ts(25,15): error TS2339: Property 'length' does not exist on type 'number'.
 tests/cases/compiler/jsdocInTypeScript.ts(30,3): error TS2339: Property 'x' does not exist on type '{}'.
+tests/cases/compiler/jsdocInTypeScript.ts(42,12): error TS2503: Cannot find namespace 'N'.
 
 
-==== tests/cases/compiler/jsdocInTypeScript.ts (5 errors) ====
+==== tests/cases/compiler/jsdocInTypeScript.ts (6 errors) ====
     // JSDoc typedef tags are not bound TypeScript files.
     /** @typedef {function} T */
     declare const x: T;
@@ -55,4 +56,9 @@ tests/cases/compiler/jsdocInTypeScript.ts(30,3): error TS2339: Property 'x' does
     function tem<T extends number>(t: T): I<T> { return {}; }
     
     let i: I; // Should succeed thanks to type parameter default
+    
+    /** @typedef {string} N.Str */
+    import M = N; // Error: @typedef does not create namespaces in TypeScript code.
+               ~
+!!! error TS2503: Cannot find namespace 'N'.
     

--- a/tests/baselines/reference/jsdocInTypeScript.js
+++ b/tests/baselines/reference/jsdocInTypeScript.js
@@ -39,6 +39,9 @@ function tem<T extends number>(t: T): I<T> { return {}; }
 
 let i: I; // Should succeed thanks to type parameter default
 
+/** @typedef {string} N.Str */
+import M = N; // Error: @typedef does not create namespaces in TypeScript code.
+
 
 //// [jsdocInTypeScript.js]
 var T = (function () {
@@ -63,3 +66,5 @@ z.x = 1; // Error
 /** @template T */
 function tem(t) { return {}; }
 var i; // Should succeed thanks to type parameter default
+/** @typedef {string} N.Str */
+var M = N; // Error: @typedef does not create namespaces in TypeScript code.

--- a/tests/cases/compiler/jsdocInTypeScript.ts
+++ b/tests/cases/compiler/jsdocInTypeScript.ts
@@ -37,3 +37,6 @@ interface I<T extends number = 0> {}
 function tem<T extends number>(t: T): I<T> { return {}; }
 
 let i: I; // Should succeed thanks to type parameter default
+
+/** @typedef {string} N.Str */
+import M = N; // Error: @typedef does not create namespaces in TypeScript code.

